### PR TITLE
Enhance mean module to process HPD matrices

### PR DESCRIPTION
**Pull request by qbarthelemy**
Use actually `_check_mean` in `mean_covariance`;
Enhance functions `mean_ale` and `mean_alm`, removing np.diag(np.diag());
Enhance all means to process HPD matrices (excepted `mean_ale`, blocked by #238), related to #204;
Complete doc and tests;
Simplify code for medians.

### CODE REVIEW

File: pyriemann/utils/mean.py [hunk 1]
```diff
@@ -14,12 +14,12 @@
 def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
```
Comment (agramfort) [inline pyriemann/utils/mean.py:14]: to go all the way covmats param could be renamed but that's a deprecation cycle whichi is never fun...

File: pyriemann/utils/mean.py [hunk 2]
```diff
@@ -30,52 +30,60 @@ def mean_ale(covmats, tol=10e-7, maxiter=50, sample_weight=None):
...
-    r"""Ando-Li-Mathias (ALM) mean of SPD matrices.
+    r"""Ando-Li-Mathias (ALM) mean of SPD/HPD matrices.
 
     Return the geometric mean recursively [1]_, generalizing from:
 
     .. math::
-        \mathbf{C} = A^{\frac{1}{2}}(A^{-\frac{1}{2}}B^{\frac{1}{2}}
-                     A^{-\frac{1}{2}})^{\frac{1}{2}}A^{\frac{1}{2}}
+        \mathbf{C} = X_1^{\frac{1}{2}} (X_1^{-\frac{1}{2}}X_2^{\frac{1}{2}}
+                     X_1^{-\frac{1}{2}})^{\frac{1}{2}} X_1^{\frac{1}{2}}
```
Comment (agramfort) [inline pyriemann/utils/mean.py:86]: we should make consistent how we document the mean equations. Here you only have 2 matrices but one can have n_matrices > 2.

**Review by agramfort — APPROVED** [review]
feel free to take my comments into account in a follow up PR

thx @qbarthelemy

File: pyriemann/utils/mean.py (hunk 1 above)
Comment (qbarthelemy) [reply pyriemann/utils/mean.py:14]: Yes, my sweetest dream is to rename all `covmats` into `X` !
We could plan to group all deprecations in a release. What do you think? @agramfort @sylvchev

File: pyriemann/utils/mean.py (hunk 2 above)
Comment (qbarthelemy) [reply pyriemann/utils/mean.py:86]: I am afraid that recursive nature of this mean prevents to write an analytic formula.
@sylvchev , what do you think?

Comment (qbarthelemy) [reply pyriemann/utils/mean.py:86]: By the way, the link http://bezout.dm.unipi.it/software/mmtoolbox/ does not exist.

Comment (sylvchev) [reply pyriemann/utils/mean.py:87]: The ALM is useful for a sanity check as it is a reference in geometry but this is not scale due to the recursion, so this is a special case.

**Review by sylvchev — APPROVED** [review]

**Comment by sylvchev** [comment]
Thanks @qbarthelemy !

File: pyriemann/utils/mean.py (hunk 1 above)
Comment (qbarthelemy) [reply pyriemann/utils/mean.py:14]: @agramfort , what is the best way to do this deprecation for module `utils.mean`?
Renaming argument `covmats` into `X`?
Renaming argument and transforming it into a positional-only argument?

Comment (agramfort) [reply pyriemann/utils/mean.py:14]: example with new signature:

def mean_ale(X, tol=10e-7, maxiter=50, sample_weight=None, covmats=None):

then if covmats is not None you raise a deprecation warning that covmats has been renamed to X and that this break in the next release.

clear?

Comment (qbarthelemy) [reply pyriemann/utils/mean.py:14]: Yes, it's clear and elegant. Thx!

**Pull request merged**